core/solver/multigrid.cpp

```diff
@@ -225,7 +228,8 @@ struct MultigridState {
 
     void run_cycle(multigrid::cycle cycle, size_type level,
                    const std::shared_ptr<const LinOp>& matrix, const LinOp* b,
-                   LinOp* x, bool is_first = true, bool is_end = true)
+                   LinOp* x, bool x_is_zero = false, bool is_first = true,
+                   bool is_end = true)
```

Comment:
This seems quite hard to use with so many boolean options (and easy to use it the wrong way because you remembered a different order).
Maybe an enum could help here (combining options with binary or)?

Reply:
agree, though it is simpler to maintain if you have separate enum parameters for the different options.